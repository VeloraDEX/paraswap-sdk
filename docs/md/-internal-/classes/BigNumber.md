[**@paraswap/sdk**](../../README.md) • **Docs**

***

[@paraswap/sdk](../../globals.md) / [\<internal\>](../README.md) / BigNumber

# Class: BigNumber

## Implements

- [`Hexable`](../interfaces/Hexable.md)

## Constructors

### new BigNumber()

> **new BigNumber**(`constructorGuard`, `hex`): [`BigNumber`](BigNumber.md)

#### Parameters

• **constructorGuard**: `any`

• **hex**: `string`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:7

## Properties

### \_hex

> `readonly` **\_hex**: `string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:5

***

### \_isBigNumber

> `readonly` **\_isBigNumber**: `boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:6

## Methods

### abs()

> **abs**(): [`BigNumber`](BigNumber.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:10

***

### add()

> **add**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:11

***

### and()

> **and**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:17

***

### div()

> **div**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:13

***

### eq()

> **eq**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:23

***

### fromTwos()

> **fromTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:8

***

### gt()

> **gt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:26

***

### gte()

> **gte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:27

***

### isNegative()

> **isNegative**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:28

***

### isZero()

> **isZero**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:29

***

### lt()

> **lt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:24

***

### lte()

> **lte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:25

***

### mask()

> **mask**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:20

***

### mod()

> **mod**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:15

***

### mul()

> **mul**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:14

***

### or()

> **or**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:18

***

### pow()

> **pow**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:16

***

### shl()

> **shl**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:21

***

### shr()

> **shr**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:22

***

### sub()

> **sub**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:12

***

### toBigInt()

> **toBigInt**(): `bigint`

#### Returns

`bigint`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:31

***

### toHexString()

> **toHexString**(): `string`

#### Returns

`string`

#### Implementation of

[`Hexable`](../interfaces/Hexable.md).[`toHexString`](../interfaces/Hexable.md#tohexstring)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:33

***

### toJSON()

> **toJSON**(`key`?): `any`

#### Parameters

• **key?**: `string`

#### Returns

`any`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:34

***

### toNumber()

> **toNumber**(): `number`

#### Returns

`number`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:30

***

### toString()

> **toString**(): `string`

Returns a string representation of an object.

#### Returns

`string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:32

***

### toTwos()

> **toTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:9

***

### xor()

> **xor**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:19

***

### from()

> `static` **from**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `any`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:35

***

### isBigNumber()

> `static` **isBigNumber**(`value`): `value is BigNumber`

#### Parameters

• **value**: `any`

#### Returns

`value is BigNumber`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:36
