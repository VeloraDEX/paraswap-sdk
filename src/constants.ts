import { EnumerateLiteral, APIVersion } from './types';

import type { ContractMethod } from '@paraswap/core';

export {
  SwapSide,
  ContractMethod,
  ContractMethodV5,
  ContractMethodV6,
} from '@paraswap/core';

// allows to pass by name instead of importing enum first
export type ContractMethodByName = EnumerateLiteral<typeof ContractMethod>;

export const API_URL = 'https://api.paraswap.io';
export const DEFAULT_VERSION = '6.2' satisfies APIVersion;
