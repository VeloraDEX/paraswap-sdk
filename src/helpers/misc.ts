import type { Primitive } from 'ts-essentials';

export const objectToFilledEntries = <T extends Record<string, unknown>>(
  object: T
): [string, string][] => {
  return (
    Object.entries(object)
      // removes keys with undefined values
      .filter(([, value]) => value !== undefined)
      .map(([key, value]) => [key, String(value)])
  );
};

export const constructSearchString = <
  //                               not arrays or mappings
  U extends Record<string, Exclude<Primitive, symbol>>
>(
  queryOptions: U
): `?${string}` | '' => {
  const queryEntries = objectToFilledEntries(queryOptions);

  const queryString = new URLSearchParams(queryEntries).toString();

  // returns empty string or `?${string}`
  return queryString && `?${queryString}`;
};

type MinAxiosError = {
  code?: string;
  request?: any;
  isAxiosError: boolean;
  message: string;
};

type MinAxiosResponse = {
  data: any;
  status: number;
  statusText: string;
  headers: Record<string, any>;
  config: { url?: string; method?: string };
};

type FetcherErrorConstructorInput = MinAxiosError & {
  response?: MinAxiosResponse;
};

export interface FetcherErrorInterface extends FetcherErrorConstructorInput {
  status?: number;
  code?: string;
}

export const isFetcherError = (error: unknown): error is FetcherError => {
  return error instanceof FetcherError;
};

export class FetcherError extends Error implements FetcherErrorInterface {
  code;
  status;
  request;
  response;
  isAxiosError = false;

  public constructor({
    code,
    request,
    response,
    isAxiosError,
    message,
  }: FetcherErrorConstructorInput) {
    super();

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, FetcherError);
    }

    this.name = this.constructor.name;
    this.code = code;
    this.request = request;
    this.isAxiosError = isAxiosError;

    // no response on AxiosError
    if (!response) {
      this.message = message;
      return;
    }

    this.response = response;

    const { data, status } = response;
    this.status = status;
    this.message = isDataWithError(data) ? data.error : message;
  }
}
// to turn `object -> Record` for indexed variable access
function isObject(obj: unknown): obj is Record<string | symbol, any> {
  return !!obj && typeof obj === 'object';
}

export function isDataWithError(data: unknown): data is { error: string } {
  return isObject(data) && typeof data['error'] === 'string';
}

export type ExtractAbiMethodNames<T extends readonly { name: string }[]> =
  T[number]['name'];

// reduce element[] to Object{key: prop, val?: element}
// for example
// gatherObjectsByProp(Token[], token => token.address) => Record<address, Token|undefined>
export function gatherObjectsByProp<T>(
  elements: T[],
  pickProp: (elem: T, index: number) => string
): Record<string, T>;
export function gatherObjectsByProp<T, U>(
  elements: T[],
  pickProp: (elem: T, index: number) => string,
  transfrom: (elem: T, accumElem: U | undefined, index: number) => U
): Record<string, U>;
export function gatherObjectsByProp<T, U>(
  elements: T[],
  pickProp: (elem: T, index: number) => string,
  transform?: (elem: T, accumElem: U | undefined, index: number) => U
): Record<string, T> | Record<string, U> {
  return elements.reduce<Record<string, T> | Record<string, U>>(
    (accum, element, index) => {
      const key = pickProp(element, index);

      const accumElem: T | U | undefined = accum[key];
      const transformedElement = transform
        ? //                       if transform is available, can only be U | undefined
          transform(element, accumElem as U | undefined, index)
        : element;

      accum[key] = transformedElement;

      return accum;
    },
    {}
  );
}

// checks that array has at least one element
export const isFilledArray = <T>(array: T[]): array is [T, ...T[]] => {
  return array.length > 0;
};

export function getRandomInt(): number {
  return Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
}

export function runOnceAndCache<T, Args extends any[]>(
  // can pass `(...any[]) => any but null | undefined`
  func: (...args: Args) => NonNullable<T>
): (...args: Args) => NonNullable<T> {
  let result: NonNullable<T>;

  return (...args) => {
    return result ?? (result = func(...args));
  };
}

export function deriveCompactSignature(signature: string): string {
  // Remove "0x" prefix if present
  if (signature.startsWith('0x')) {
    signature = signature.slice(2);
  }

  // Convert the hex string to a byte array
  const bytes = new Uint8Array(signature.length / 2);
  for (let i = 0; i < signature.length; i += 2) {
    bytes[i / 2] = parseInt(signature.slice(i, i + 2), 16);
  }

  // Validate the signature length (64 or 65 bytes)
  if (bytes.length !== 64 && bytes.length !== 65) {
    throw new Error('Invalid signature length: must be 64 or 65 bytes');
  }

  // Extract r and s components
  const r = `0x${Array.from(bytes.slice(0, 32), (b) =>
    b.toString(16).padStart(2, '0')
  ).join('')}`;
  let v;

  // Handle 64-byte (EIP-2098 compact) and 65-byte signatures
  if (bytes.length === 64) {
    // Extract v from the highest bit of s and clear the bit in s
    v = 27 + (bytes[32]! >> 7);
    bytes[32]! &= 0x7f; // Clear the highest bit
  } else {
    // Extract v directly for 65-byte signature
    v = bytes[64]!;

    // Normalize v to canonical form (27 or 28)
    if (v < 27) {
      v += 27;
    }
  }

  // Compute yParityAndS (_vs) for the compact signature
  const sBytes = Array.from(bytes.slice(32, 64));
  if (v === 28) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    sBytes[0]! |= 0x80; // Set the highest bit if v is 28
  }
  const yParityAndS = `0x${sBytes
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')}`;

  // Construct the compact signature by concatenating r and yParityAndS
  const compactSignature = r + yParityAndS.slice(2);

  return compactSignature;
}
