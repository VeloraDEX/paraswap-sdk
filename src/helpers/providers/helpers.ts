import { assert } from 'ts-essentials';
import { TypedDataField } from '../../methods/common/orders/buildOrderData';

// regex from @ethersproject/hash TypedDataEncoder.constructor
// may be overly strict, but reliable
const baseTypeRegex = /^([^\x5b]*)(\x5b|$)/;

export function findPrimaryType(
  types: Record<string, TypedDataField[]>
): string {
  const candidates = Object.keys(types);
  const candidatesSet = new Set(candidates);

  candidates.forEach((candidate) => {
    const typedDataFields = types[candidate];
    if (!typedDataFields) return;

    typedDataFields.forEach(({ type }) => {
      // Get the base type (drop any array specifiers)
      const baseType = type.match(baseTypeRegex)?.[1];
      if (!baseType) return;

      // if type was referred to as a child of another type, it can't be the primaryType
      candidatesSet.delete(baseType);
    });
  });

  const [primaryType] = Array.from(candidatesSet);

  assert(
    primaryType,
    `No primary type found in SignableTypedData types, ${JSON.stringify(types)}`
  );

  return primaryType;
}
