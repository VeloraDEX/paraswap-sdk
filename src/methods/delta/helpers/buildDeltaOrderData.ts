import { MarkOptional } from 'ts-essentials';
import { Domain, ZERO_ADDRESS } from '../../common/orders/buildOrderData';
import { DeltaAuctionOrder } from './types';
import { composeDeltaOrderPermit } from './composePermit';
import { DeltaPrice } from '../getDeltaPrice';

// Order(address owner,address beneficiary,address srcToken,address destToken,uint256 srcAmount,uint256 destAmount,uint256 deadline,uint256 nonce,bytes permit)";
const Order = [
  { name: 'owner', type: 'address' },
  { name: 'beneficiary', type: 'address' },
  { name: 'srcToken', type: 'address' },
  { name: 'destToken', type: 'address' },
  { name: 'srcAmount', type: 'uint256' },
  { name: 'destAmount', type: 'uint256' },
  { name: 'expectedDestAmount', type: 'uint256' },
  { name: 'deadline', type: 'uint256' },
  { name: 'nonce', type: 'uint256' },
  { name: 'partnerAndFee', type: 'uint256' },
  { name: 'permit', type: 'bytes' },
];

export type SignableDeltaOrderData = {
  types: {
    Order: typeof Order;
  };
  domain: Domain;
  data: DeltaAuctionOrder;
};

type SignDeltaOrderInput = {
  orderInput: DeltaAuctionOrder;
  paraswapDeltaAddress: string;
  chainId: number;
};

function produceDeltaOrderTypedData({
  orderInput,
  chainId,
  paraswapDeltaAddress,
}: SignDeltaOrderInput): SignableDeltaOrderData {
  const typedData = {
    types: { Order },
    domain: {
      name: 'Portikus',
      version: '2.0.0',
      chainId,
      verifyingContract: paraswapDeltaAddress,
    },
    data: orderInput,
  };

  return typedData;
}

export type DeltaOrderDataInput = MarkOptional<
  Omit<DeltaAuctionOrder, 'partnerAndFee'>,
  'beneficiary' | 'deadline' | 'nonce' | 'permit'
> &
  Pick<DeltaPrice, 'partnerFee'>;

export type BuildDeltaOrderDataInput = DeltaOrderDataInput & {
  partnerAddress: string;
  paraswapDeltaAddress: string;
  takeSurplus: boolean;
  chainId: number;
};

// default deadline = 1 hour from now (may be changed later)
export const DELTA_DEFAULT_EXPIRY = 60 * 60; // seconds

export function buildDeltaSignableOrderData({
  owner,
  beneficiary = owner,

  srcToken,
  destToken,
  srcAmount,
  destAmount,
  expectedDestAmount,

  deadline = Math.floor(Date.now() / 1000 + DELTA_DEFAULT_EXPIRY),
  nonce = Date.now().toString(10), // random enough to not cause collisions

  permit = '0x',

  partnerAddress,
  partnerFee,
  takeSurplus,

  chainId,
  paraswapDeltaAddress,
}: BuildDeltaOrderDataInput): SignableDeltaOrderData {
  const orderInput: DeltaAuctionOrder = {
    owner,
    beneficiary,
    srcToken,
    destToken,
    srcAmount,
    destAmount,
    expectedDestAmount,
    deadline,
    nonce,
    permit: composeDeltaOrderPermit({ permit, nonce }),
    partnerAndFee: producePartnerAndFee({
      partnerFee,
      partnerAddress,
      takeSurplus,
    }),
  };

  return produceDeltaOrderTypedData({
    orderInput,
    chainId,
    paraswapDeltaAddress,
  });
}

type ProducePartnerAndFeeInput = {
  partnerFee: number;
  partnerAddress: string;
  takeSurplus: boolean;
};

// fee and address are encoded together
function producePartnerAndFee({
  partnerFee,
  partnerAddress,
  takeSurplus,
}: ProducePartnerAndFeeInput): string {
  if (partnerAddress === ZERO_ADDRESS) return '0';

  const partnerFeeBps = BigInt((partnerFee * 100).toFixed(0));
  const partnerAndFee =
    (BigInt(partnerAddress) << BigInt(96)) |
    partnerFeeBps |
    (BigInt(takeSurplus) << BigInt(8));

  return partnerAndFee.toString(10);
}
