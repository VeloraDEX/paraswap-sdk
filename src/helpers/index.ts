import { EthersV5ProviderDeps } from './providers/ethers';
import { EthersV6ProviderDeps } from './providers/ethersV6';
export { constructFetcher as constructAxiosFetcher } from './fetchers/axios';
export { constructFetcher as constructFetchFetcher } from './fetchers/fetch';
export {
  constructEthersV5ContractCaller as constructEthersContractCaller,
  constructEthersV5ContractCaller,
  EthersV5ProviderDeps,
} from './providers/ethers';
export {
  constructContractCaller as constructEthersV6ContractCaller,
  EthersV6ProviderDeps,
} from './providers/ethersV6';
export type EthersProviderDeps = EthersV5ProviderDeps | EthersV6ProviderDeps;

export {
  constructContractCaller as constructWeb3ContractCaller,
  Web3UnpromiEvent,
} from './providers/web3';
export {
  constructContractCaller as constructViemContractCaller,
  MinViemClient,
  txParamsToViemTxParams,
} from './providers/viem';
export { constructToken } from './token';
export { isFetcherError } from './misc';
export type { FetcherError, FetcherErrorInterface } from './misc';
