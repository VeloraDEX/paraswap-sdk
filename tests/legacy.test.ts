import * as dotenv from 'dotenv';
import Web3 from 'web3';
import { ethers } from 'ethersV5';
import fetch from 'isomorphic-unfetch';
import {
  ParaSwap,
  Token,
  Allowance,
  TransactionParams,
  SwapSide,
  OptimalRate,
} from '../src';
import BigNumber from 'bignumber.js';
import { APIError } from '../src/legacy';
import erc20abi from './abi/ERC20.json';

import { assert } from 'ts-essentials';
import { HardhatProvider, setupFork } from './helpers/hardhat';

dotenv.config();

jest.setTimeout(30 * 1000);

const ETH = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
const BAT = '0x0d8775f648430679a709e98d2b0cb6250d2887ef';
const MANA = '0x0f5d2fb29fb7d3cfee444a200298f468908cc942';
const chainId = 1;
const srcToken = ETH;
const destToken = DAI;
const srcAmount = (1 * 1e18).toString(); //The source amount multiplied by its decimals

const referrer = 'sdk-test';

const wallet = ethers.Wallet.createRandom();

const web3Provider = new Web3(HardhatProvider as any);

const ethersProvider = new ethers.providers.Web3Provider(
  HardhatProvider as any
);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address;

describe('ParaSwap SDK', () => {
  let paraSwap: ParaSwap;

  beforeAll(async () => {
    await setupFork({
      accounts: [{ address: senderAddress, balance: 8e18 }],
    });
    paraSwap = new ParaSwap({ chainId, fetch, version: '5' }).setWeb3Provider(
      web3Provider
    );
  });

  test('Get_Balance', async () => {
    const balance = await paraSwap.getBalance(senderAddress, ETH);
    expect(balance).toBeDefined();
  });

  test('Get_Markets', async () => {
    const markets = await paraSwap.getMarketNames();
    expect((markets as string[]).length).toBeGreaterThan(15);
  });

  test('Get_Tokens', async () => {
    const tokensOrError = await paraSwap.getTokens();

    const tokens = tokensOrError as Token[];

    expect(Array.isArray(tokens)).toBe(true);
    expect(tokens.length).toBeGreaterThan(0);
    expect(tokens[0]).toEqual(
      expect.objectContaining({
        symbol: expect.any(String),
        address: expect.any(String),
        decimals: expect.any(Number),
      })
    );
  });

  test('Get_Rates', async () => {
    const ratesOrError = await paraSwap.getRate(
      'ETH',
      'DAI',
      srcAmount,
      senderAddress,
      SwapSide.SELL,
      { includeDEXS: ['UniswapV2'], otherExchangePrices: true }
    );
    const priceRoute = ratesOrError as OptimalRate;

    const { destAmount, bestRoute, others } = priceRoute;

    expect(typeof destAmount).toBe('string');

    expect(Array.isArray(bestRoute)).toBe(true);

    const swapExchange = bestRoute[0]?.swaps[0]?.swapExchanges[0];

    assert(swapExchange, 'exchange available at swapExchanges[0]');

    expect(typeof swapExchange.destAmount).toBe('string');
    expect(new BigNumber(swapExchange.destAmount).isNaN()).toBe(false);

    expect(typeof swapExchange.exchange).toBe('string');

    const firstBestRoute = bestRoute[0];
    assert(firstBestRoute, 'route available at bestRoute[0]');

    expect(typeof firstBestRoute.percent).toBe('number');
    expect(new BigNumber(firstBestRoute.percent).isNaN()).toBe(false);

    expect(typeof swapExchange.srcAmount).toBe('string');
    expect(new BigNumber(swapExchange.srcAmount).isNaN()).toBe(false);

    expect(Array.isArray(others)).toBe(true);

    const firstRoute = others?.[0];

    assert(firstRoute, 'at least one route must exist');

    expect(typeof firstRoute.exchange).toBe('string');

    expect(typeof firstRoute.unit).toBe('string');
    expect(new BigNumber(firstRoute.unit as string).isNaN()).toBe(false);
  });

  test('Get_Spender', async () => {
    const spender = await paraSwap.getTokenTransferProxy();
    expect(web3Provider.utils.isAddress(spender as string));
  });

  test('Get_Allowance', async () => {
    const allowance = await paraSwap.getAllowance(senderAddress, DAI);

    if (!allowance || (allowance as APIError).message) {
      return;
    }

    expect(new BigNumber((allowance as Allowance).allowance).isNaN()).toBe(
      false
    );
  });

  test('Get_Allowances', async () => {
    const allowancesOrError = await paraSwap.getAllowances(senderAddress, [
      DAI,
      BAT,
      MANA,
    ]);

    if ('message' in allowancesOrError) {
      return;
    }

    const allowances = allowancesOrError;

    allowances.forEach((allowance) =>
      expect(new BigNumber(allowance.allowance).isNaN()).toBe(false)
    );
  });

  test('Get_Adapters', async () => {
    const adaptersOrError = await paraSwap.getAdapters();

    const adapters = adaptersOrError as string[];

    expect(adapters).toMatchSnapshot('Get_Adapters');
  });

  test('Build_Tx', async () => {
    const destToken = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
    const ratesOrError = await paraSwap.getRate(
      srcToken,
      destToken,
      srcAmount,
      senderAddress,
      SwapSide.SELL,
      { includeDEXS: ['UniswapV2'] }
    );

    expect((ratesOrError as APIError)?.data?.error).toBeUndefined();

    const priceRoute = ratesOrError as OptimalRate;

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txOrError = await paraSwap.buildTx(
      srcToken,
      destToken,
      srcAmount,
      destAmount,
      priceRoute,
      senderAddress,
      referrer,
      undefined,
      undefined,
      undefined,
      { ignoreChecks: true }
    );

    expect(txOrError.data.error).toBeUndefined();
    expect(typeof txOrError).toBe('object');
  });

  test('Build_and_Send_Tx_SELL', async () => {
    const ratesOrError = await paraSwap.getRate(
      srcToken,
      destToken,
      srcAmount,
      senderAddress,
      SwapSide.SELL,
      { includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'] }
    );
    const priceRoute = ratesOrError as OptimalRate;
    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txOrError = await paraSwap.buildTx(
      srcToken,
      destToken,
      srcAmount,
      destAmount,
      priceRoute,
      signer.address,
      referrer,
      undefined,
      undefined,
      undefined,
      { ignoreChecks: true }
    );
    const _tx = txOrError as TransactionParams;
    const transaction = {
      ..._tx,
      gasPrice: _tx.gasPrice && '0x' + new BigNumber(_tx.gasPrice).toString(16),
      maxFeePerGas:
        _tx.maxFeePerGas && '0x' + new BigNumber(_tx.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        _tx.maxPriorityFeePerGas &&
        '0x' + new BigNumber(_tx.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(_tx.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
  test('Build_and_Send_Tx_BUY', async () => {
    const destAmount = srcAmount;
    const ratesOrError = await paraSwap.getRate(
      srcToken,
      destToken,
      destAmount,
      senderAddress,
      SwapSide.BUY,
      { includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'] }
    );
    const priceRoute = ratesOrError as OptimalRate;
    const _srcAmount = new BigNumber(priceRoute.srcAmount)
      .times(1.1)
      .toFixed(0);

    const txOrError = await paraSwap.buildTx(
      srcToken,
      destToken,
      _srcAmount,
      destAmount,
      priceRoute,
      signer.address,
      referrer,
      undefined,
      undefined,
      undefined,
      { ignoreChecks: true }
    );
    const _tx = txOrError as TransactionParams;
    const transaction = {
      ..._tx,
      gasPrice: _tx.gasPrice && '0x' + new BigNumber(_tx.gasPrice).toString(16),
      maxFeePerGas:
        _tx.maxFeePerGas && '0x' + new BigNumber(_tx.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        _tx.maxPriorityFeePerGas &&
        '0x' + new BigNumber(_tx.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(_tx.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
});
