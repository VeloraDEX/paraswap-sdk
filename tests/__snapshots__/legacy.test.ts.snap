// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParaSwap SDK Get_Adapters: Get_Adapters 1`] = `
[
  "Bancor",
  "Compound",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "CurveV2",
  "Lido",
  "EtherFi",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "PancakeSwapV2",
  "SushiSwap",
  "ShibaSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "PancakeswapV3",
  "AaveV2",
  "AaveV3",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "MakerPsm",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "CurveV1StableNg",
  "wstETH",
  "Hashflow",
  "MaverickV1",
  "SwaapV2",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Wombat",
  "Swell",
  "Spark",
  "sUSDS",
  "AugustusRFQ",
]
`;
